src/include/duckdb/common/types.hpp

```diff
@@ -486,10 +496,28 @@ struct StringType {
 
 struct ListType {
 	DUCKDB_API static const LogicalType &GetChildType(const LogicalType &type);
+	// Non-constant on the logical type so we can bind it in case is a user type
+	DUCKDB_API static LogicalType &GetChildTypeBind( LogicalType &type);
+};
+
+struct UserType{
+	DUCKDB_API static const string &GetTypeName(const LogicalType &type);
 };
 
+struct EnumType{
+	DUCKDB_API static const string &GetTypeName(const LogicalType &type);
+	DUCKDB_API static int64_t GetPos(const LogicalType &type, const string& key);
+	DUCKDB_API static const vector<string> &GetValuesInsertOrder(const LogicalType &type);
+	DUCKDB_API static idx_t GetSize(const LogicalType &type);
+	DUCKDB_API static shared_ptr<ExtraTypeInfo> CreateEnumInfo(const string & enum_name, const shared_ptr<vector<string>>& ordered_data);
```

Comment:
Does this need to be a separate function? It seems that this is only used like this:

```cpp
LogicalType::ENUM(EnumType::CreateEnumInfo(name, array), array->size()));
```

Can't we turn that into this?
```cpp
LogicalType::ENUM(name, array, array->size()));
```

Or even
```cpp
LogicalType::ENUM(name, array));
```

Also, do we need to use a `shared_ptr` to a vector of string here? Remember that the ExtraTypeInfo itself is already a `shared_ptr`.